pkg/scorecard.go

```diff
@@ -277,3 +282,131 @@ func ExperimentalRunProbes(ctx context.Context,
 		projectClient,
 	)
 }
+
+type runConfig struct {
+	client        clients.RepoClient
+	vulnClient    clients.VulnerabilitiesClient
+	ciiClient     clients.CIIBestPracticesClient
+	projectClient packageclient.ProjectPackageClient
+	ossfuzzClient clients.RepoClient
+	checks        []string
+	commit        string
+	probes        []string
+	commitDepth   int
+}
+
+type Option func(*runConfig) error
+
+func WithCommitDepth(depth int) Option {
+	return func(c *runConfig) error {
+		c.commitDepth = depth
+		return nil
+	}
+}
+
+func WithCommitSHA(sha string) Option {
+	return func(c *runConfig) error {
+		c.commit = sha
+		return nil
+	}
+}
+
+func WithChecks(checks []string) Option {
+	return func(c *runConfig) error {
+		c.checks = checks
+		return nil
+	}
+}
+
+func WithProbes(probes []string) Option {
+	return func(c *runConfig) error {
+		c.probes = probes
+		return nil
+	}
+}
+
+func WithRepoClient(client clients.RepoClient) Option {
+	return func(c *runConfig) error {
+		c.client = client
+		return nil
+	}
+}
+
+func WithOSSFuzzClient(client clients.RepoClient) Option {
+	return func(c *runConfig) error {
+		c.ossfuzzClient = client
+		return nil
+	}
+}
+
+func WithVulnerabilitiesClient(client clients.VulnerabilitiesClient) Option {
+	return func(c *runConfig) error {
+		c.vulnClient = client
+		return nil
+	}
+}
+
+func WithOpenSSFBestPraticesClient(client clients.CIIBestPracticesClient) Option {
+	return func(c *runConfig) error {
+		c.ciiClient = client
+		return nil
+	}
+}
+
+func Run(ctx context.Context, repo clients.Repo, options ...Option) (ScorecardResult, error) {
+	// TODO logger
+	logger := sclog.NewLogger(sclog.InfoLevel)
+	c := runConfig{
+		commit: clients.HeadSHA,
+	}
+	for _, option := range options {
+		if err := option(&c); err != nil {
+			return ScorecardResult{}, err
+		}
+	}
+	if c.ciiClient == nil {
+		c.ciiClient = clients.DefaultCIIBestPracticesClient()
+	}
+	if c.ossfuzzClient == nil {
+		c.ossfuzzClient = ossfuzz.CreateOSSFuzzClient(ossfuzz.StatusURL)
+	}
+	if c.vulnClient == nil {
+		c.vulnClient = clients.DefaultVulnerabilitiesClient()
+	}
+	if c.projectClient == nil {
+		c.projectClient = packageclient.CreateDepsDevClient()
+	}
+
+	var requiredRequestTypes []checker.RequestType
+	var err error
+	switch repo.(type) {
+	case *localdir.Repo:
+		requiredRequestTypes = append(requiredRequestTypes, checker.FileBased)
+		if c.client == nil {
+			c.client = localdir.CreateLocalDirClient(ctx, logger)
+		}
+	case *githubrepo.Repo:
+		if c.client == nil {
+			c.client = githubrepo.CreateGithubRepoClient(ctx, logger)
+		}
+	case *gitlabrepo.Repo:
+		if c.client == nil {
+			c.client, err = gitlabrepo.CreateGitlabClient(ctx, repo.Host())
+			if err != nil {
+				return ScorecardResult{}, fmt.Errorf("creating gitlab client: %w", err)
+			}
+		}
+	}
+
+	if !strings.EqualFold(c.commit, clients.HeadSHA) {
```

Comment:
It appears `--repo github.com/foo/bar --commit HEAD` would have an empty slice of requiredRequestTypes in this example, is that intentional?

Reply:
`--commit HEAD` is a no-op since that's the default. This is what the existing code already did.



https://github.com/ossf/scorecard/blob/91532e12d129f4f377a785269e560fbba82b86d4/cmd/root.go#L113-L124